File: packages/models/src/Topogon.ts
import Area, { type IArea } from "./Area.js";
import Boulder, { type IBoulder } from "./Boulder.js";
import Route, { type IRoute } from "./Route.js";
import Topo, { type ITopo } from "./Topo.js";

export type ITopogon = {
  id: number;
  label?: string;
  topo?: ITopo;
  data: string; // TODO: JSON or SVG string
  // Only can have one of these
  area?: IArea;
  boulder?: IBoulder;
  route?: IRoute;
};

interface Topogon
  extends Omit<ITopogon, "topo" | "area" | "boulder" | "route"> {
  topo: Topo;
  area?: Area;
  boulder?: Boulder;
  route?: Route;
}

class Topogon {
  constructor(data: ITopogon) {
    this.id = data.id;
    this.label = data.label;
    this.data = data.data;
    if (data.topo) {
      this.topo = new Topo(data.topo);
    }

    let relations = 0;
    if (data.area) {
      this.area = Area.build(data.area);
      relations += 1;
    }
    if (data.boulder) {
      this.boulder = Boulder.build(data.boulder);
      relations += 1;
    }
    if (data.route) {
      this.route = Route.build(data.route);
      relations += 1;
    }
    if (relations > 1) {
      throw new Error("Topogon can only have one of crag, boulder, or route");
    }
  }
}

export default Topogon;

File: go-server/models/topogon.go
package models

type Topogon struct {
	ID      int      `json:"id"`
	Label   *string  `json:"label,omitzero"`
	Topo    *Topo    `json:"topo,omitzero"`
	Area    *Area    `json:"area,omitzero"`
	Boulder *Boulder `json:"boulder,omitzero"`
	Route   *Route   `json:"route,omitzero"`
}

func (t Topogon) IsZero() bool {
	return t.ID == 0 && t.Label == nil && t.Topo == nil && t.Area == nil && t.Boulder == nil && t.Route == nil
}
